src/app/modules/item/task/task-xd-pr.ts

```diff
@@ -0,0 +1,303 @@
+    this.chan.bind('platform.askHint', (hintToken : string) => platform.askHint(hintToken));
+    this.chan.bind('platform.updateDisplay', (data : TaskDisplayData) => platform.updateDisplay(data));
+    this.chan.bind('platform.openUrl', (url : string) => platform.openUrl(url));
+    this.chan.bind('platform.log', (data : TaskLog) => platform.log(data));
+
+    // Legacy calls
+    this.chan.bind('platform.updateHeight', (height : number) => platform.updateDisplay({ height: height }));
+  }
+
+  /**
+   * Task API functions
+   */
+  load(views : Object): Observable<void> {
+    return this.chan.call({
+      method: "task.load",
+      params: views,
+    });
+  }
+
+  unload() : Observable<void> {
+    return this.chan.call({
+      method: "task.unload",
+      timeout: 2000,
+    });
+  }
+
+  getHeight() : Observable<number> {
+    // TODO: validator
+    return this.chan.call({
+      method: "task.getHeight",
+      timeout: 500,
+    });
+  }
+
+  updateToken(token : string) : Observable<void> {
+    return this.chan.call({
+      method: "task.updateToken",
+      params: token,
+      timeout: 10000,
+    });
+  }
+
+  getMetaData() : Observable<TaskMetaData> {
+    // TODO: validator
+    return this.chan.call({
+      method: "task.getMetaData",
+      timeout: 2000
+    });
+  }
+
+  getAnswer() : Observable<string> {
+    // TODO: validator
+    return this.chan.call({
+      method: "task.getAnswer",
+      timeout: 2000
+    });
+  }
+
+  reloadAnswer(answer : string) : Observable<void> {
+    return this.chan.call({
+      method: "task.reloadAnswer",
+      params: answer,
+      timeout: 2000
+    });
+  }
+
+  getState() : Observable<string> {
+    // TODO: validator
+    return this.chan.call({
+      method: "task.getState",
+      timeout: 2000
+    });
+  }
+
+  reloadState(state : string) : Observable<void> {
+    return this.chan.call({
+      method: "task.reloadState",
+      params: state,
+      timeout: 2000
+    });
+  }
+
+  getViews() : Observable<TaskViews> {
+    // TODO: validator
+    return this.chan.call({
+      method: "task.getViews",
+      timeout: 2000
+    });
+  }
+
+  showViews(views : Object) : Observable<void> {
+    return this.chan.call({
+      method: "task.showViews",
+      params: views,
+      timeout: 2000
+    });
+  }
+
+  gradeAnswer(answer : string, answerToken : string) : Observable<TaskGrade> {
+    // TODO: validator
+    return this.chan.call({
+      method: "task.gradeAnswer",
+      params: [ answer, answerToken ],
+      timeout: 40000
+    });
+  }
+
+  getResources() : Observable<TaskResources> {
+    // TODO: validator
+    return this.chan.call({
+      method: "task.getResources",
+      params: [],
+      timeout: 2000
+    });
+  }
+}
+
+/*
+ * Platform object definition, created from a Task object (see below)
+ */
+
+export class Platform {
+  task : Task;
+  constructor(task : Task) {
+    this.task = task;
+  }
+
+  getTask() : Task {
+    return this.task;
+  }
```

Comment:
A priori (I don't have the full picture), I would say a listener should not have the reference of the object it is listening to. Is it used ?

Reply:
It is used by the "Platform", for instance in [platform.validate in item-display](https://github.com/France-ioi/AlgoreaFrontend/blob/1478336586c442642df72db451b669d4978ea08d/src/app/modules/item/pages/item-display/item-display.component.ts#L164), as the Platform needs to get the current answer to be validated.